ipykernel/iostream.py

```diff
@@ -314,7 +316,9 @@ def write(self, string):
                 string = string.decode(self.encoding, 'replace')
 
             is_child = (not self._is_master_process())
+            self._buffer_lock.acquire(True)
```

Comment:
To expand a bit: if the code between `acquire()` and `release()` throws an error for some reason, the lock would never be released. You can make sure that some 'after' code is run even if an exception is thrown:

```python
lock.acquire()
try:
    do_stuff()
finally:
    lock.release()
```

But locks support a more convenient way to achieve the same thing:

```python
with lock:
    do_stuff()
```